036-manual-step/proposal.md

```diff
@@ -0,0 +1,193 @@
+plans which are incorrect or harmful. Applying generated plans without manual
+review is dangerous and may result in the loss of data or service availability.
+Manual review of the plans is required to ensure that infrastructure changes
+are executed safely.
+
+For example, in Terraform one runs `terraform plan -out=path/to/potential.plan`,
+which prints a human-readable plan to `stdout` and produces the file
+`path/to/potential.plan`. A correct workflow requires pausing, reviewing the
+human-readable plan, and only after approving the plan, proceeding to run
+`terraform apply path/to/potential.plan` in order to actually execute the plan
+and make infrastructural changes.
+
+
+# Proposal
+
+This proposal advocates for the addition of a `manual` step to the existing
+variety of steps which may be used in a job (e.g. `get`, `task`, `put`). When
+a `manual` step is added to a job, the Concourse API should expose two options:
+an "approve" option, and a "reject" option. If a user should select the
+"approve" option, then the manual step should be marked as successful and the
+build should proceed normally. If a user should select the "reject" option,
+then the manual step should be marked as failed, and the build should respond
+accordingly, as if any other step in the build had failed.
+
+
+## Supervisors
+
+It is occasionally desirable for the authority to approve `manual` steps to be
+restricted to a limited subset of users. An RBAC role, `supervisor`, should be
+added to Concourse teams. A user with the `supervisor` role has the power to
+approve manual steps. Team admins, by virtue of having the power to alter
+pipelines and could remove the manual step if they so desired, also have the
+power to approve manual steps.
+
+
+## Pipeline YAML
+
+The following represents a proposed schema for adding `manual` steps to jobs in
+pipelines:
+
+```yaml
+manual:
+  name: required string
+  timeout: required duration
+  message: optional string
+  supervisor_required: optional bool
+```
+
+All manual steps must have a _name_, and all manual approval names must be
+unique within the context of a single pipeline. This restriction is in place to
+ensure that a submitted approval or rejection is directed to the appropriate
+`manual` step, in case multiple `manual` steps are open simultaneously as
+distributed by an `in_parallel` step, or across multiple jobs within the
+pipeline. Attempting to run `fly set-pipeline` on a pipeline where multiple
+`manual` steps exist with the same name should result in a validation error
+rejecting the proposed pipeline.
+
+Although the _timeout_ step modifier is usually an optional addition to steps,
+each manual step must define a timeout step modifier. Manual steps are special
+in this requirement (compared to other steps) in order to ensure that forgotten
+jobs with open manual approvals, including their task containers and volumes,
+are garbage-collected.
```

Comment:
I see what you're getting at here.

The design presumption I'm working from is, it's better to be explicit about requiring the user to set expectations, than to allow the user an easy way to shoot himself in the foot.

Let's take GitHub PRs as an example - it's pretty easy to end up with PRs staying open for indefinite periods of time, because priorities, and people switch their attention away and forget about stuff that was opened. The cost for a PR staying open for months on-end is basically zero (maybe some lost effort whose value won't be realized since it's not merged in), so it's fine not to add complexity to people opening PRs by asking them to define, when they open the PR, when to self-close the PR (even asking that is kind of hostile, from a design perspective).

But the cost for a manual step staying open is a resource leak, one that is sure to make Concourse operators tear their hair out: for one, because debugging silent resource leaks is *annoying* and difficult to pin down, and when the operators find that their users are at fault (because of laziness?), they're going to feel like there's nothing they can do to prevent it from recurring. It would make the system look broken.

Sure, the user can provide a `timeout:` step-modifier, but if it's not required by the system, then it'll pass `fly validate-pipeline`, and I don't think that's a (forgive the euphemism) healthy design. I think a better compromise is to require the user to provide `timeout:`, but allow the user to specify `timeout: 0` (for unlimited) if the user really wants it (really this should be true of all steps, but I digress).

Allowing operators to set overall step max execution time is a good idea, but tricky - one instance may have both long-running jobs and short jobs. But probably a good idea in practice.